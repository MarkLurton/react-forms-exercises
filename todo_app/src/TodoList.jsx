import React, { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import { v4 as uuid } from "uuid";

const TodoList = () => {
  const INITIAL_STATE = [
    {
      id: uuid(),
      task: "Water Plants",
      complete: false,
    },
    {
      id: uuid(),
      task: "Vacuum Apartment",
      complete: false,
    },
    {
      id: uuid(),
      task: "Make Bed",
      complete: false,
    },
  ];
  const [todos, setTodos] = useState(INITIAL_STATE);

  const addTodo = (todo) => {
    const newTodo = { ...todo };
    setTodos([...todos, newTodo]);
  };

  const removeTodo = (id) => {
    setTodos(todos.filter((todo) => todo.id !== id));
  };

  const markComplete = (id) => {
    setTodos(
      todos.map((task) => {
        if (task.id === id) {
          task.complete = true;
        }
        return task;
      })
    );
  };

  return (
    <div>
      <NewTodoForm addTodo={addTodo} />
      {todos.map((todo) => (
        <Todo
          key={todo.id}
          id={todo.id}
          task={todo.task}
          complete={todo.complete}
          markComplete={() => markComplete(todo.id)}
          removeTodo={() => removeTodo(todo.id)}
        />
      ))}
    </div>
  );
};

export default TodoList;
