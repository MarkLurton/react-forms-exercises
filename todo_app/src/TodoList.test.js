import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add new todo", function () {
  const { queryByText, queryByLabelText, container } = render(<TodoList />);
  const input = queryByLabelText("Task");
  const btn = queryByText("Add Todo");
  let boxes = container.getElementsByClassName("Todo");
  expect(boxes.length).toBe(3);
  fireEvent.change(input, { target: { value: "Do Dishes" } });
  fireEvent.click(btn);
  boxes = container.getElementsByClassName("Todo");
  let todo = queryByText("Do Dishes");
  expect(boxes.length).toBe(4);
  expect(todo).toBeInTheDocument();
});

it("should remove item", function () {
  const { queryAllByText, container } = render(<TodoList />);
  let todos = container.getElementsByClassName("Todo");
  expect(todos.length).toBe(3);
  const btn = queryAllByText("X");
  fireEvent.click(btn[0]);
  todos = container.getElementsByClassName("Todo");
  expect(todos.length).toBe(2);
});
