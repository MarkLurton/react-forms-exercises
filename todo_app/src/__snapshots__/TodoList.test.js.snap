// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="task"
      >
        Task
      </label>
      <input
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button>
        Add Todo
      </button>
    </form>
    <div
      class="Todo"
    >
      <p>
        Water Plants
      </p>
      <button>
        Mark as completed
      </button>
      <button>
        X
      </button>
    </div>
    <div
      class="Todo"
    >
      <p>
        Vacuum Apartment
      </p>
      <button>
        Mark as completed
      </button>
      <button>
        X
      </button>
    </div>
    <div
      class="Todo"
    >
      <p>
        Make Bed
      </p>
      <button>
        Mark as completed
      </button>
      <button>
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
